Continental geotherm initial temperature condition

anne-glerum opened the pull request:
This sets up a laterally homogeneous continental geotherm as initial temperature condition for for example rift models.

naliboff commented on include/aspect/initial_temperature/continental_geotherm.h:
```diff
@@ -0,0 +1,119 @@
+/*
+  Copyright (C) 2014 - 2016 by the authors of the ASPECT code.
```
Update copyright

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
update copyright

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+      AssertThrow(std::find(heating_models.begin(), heating_models.end(), "compositional heating") != heating_models.end(),
+                  ExcMessage("The continental geotherm initial temperature plugin requires the compositional heating plugin."));
+
+      // Check that the required material model ("visco plastic") is used
```
Would it also make sense to use this with `DiffusionDislocation` or any other material model?

anne-glerum replied:
That would make sense, any material model that provides the density and diffusivity for 3 compositions could be used. However, by restricting it to `visco plastic`, I know which parameters to read in `parse_parameters`. How would I otherwise decide which ones to read there?

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+    temperature (const double depth) const
+    {
+      // Compute some constants to calculate the temperatures T1 and T2 at the interfaces
+      // between the layers upper crust/lower crust and lower crust/lithospheric mantle.
```
Are these specific constants referred to in Turcotte and Schubert? If so, would it be possible to reference the equations?

naliboff replied:
Same for the equations below. I find these a bit harder to read/interpret than the Chapman equations, so it would be nice to have some specific references (if possible).

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+
+      //Temperature at boundary between layer 1 and 2
+      const double T1 = (a*b + conductivities[1]/thicknesses[1]*c*d*b) / (1.-(conductivities[1]*conductivities[1])/(thicknesses[1]*thicknesses[1])*d*b);
+      //Temperature at boundary between layer 2 and 3
```
Formatting: add space between `//` and`Temperature`

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+      // Temperature in layer 3
+      else if (depth <= thicknesses[0]+thicknesses[1]+thicknesses[2])
+        return (LAB_isotherm-T2)/thicknesses[2] *(depth-thicknesses[0]-thicknesses[1]) + T2;
+      // Return a constant sublithospheric temperature of 10*LAB_isotherm.
```
Will a factor  10 always be sufficient to exceed the adiabatic temperature? I imagine so, but perhaps 100 to be safe?

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+      {
+        prm.enter_subsection("Continental geotherm");
+        {
+          prm.declare_entry ("Layer thicknesses", "30000.",
```
Will this only work if you have 3 layers within the lithosphere? If this is the case, can you state this in the documentation below and add an exception in the parse parameter section?

anne-glerum replied:
This will only work for 3 layers. I've added that to the documentation. There is no need to make an additional exception, because the `possibly_extend_from_1_to_N ` function will throw for any other length than 1 or 3.

naliboff replied:
What happens if only one layer thickness is specified? It looks like there will be an error when the code tries to access `thicknesses[1]` and `thicknesses[2]`.

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+
+
+      // Retrieve the indices of the fields that represent the lithospheric layers.
+      AssertThrow(this->introspection().compositional_name_exists("upper"),ExcMessage("We need a compositional field called 'upper' representing the upper crust."));
```
Hmm, enforcing the names to be `"upper"` and `"lower"` is really specific, but I know this was done to be consistent with the continental extension cookbook. Could we allow the name to be `upper` or `upper_crust`, etc?

anne-glerum replied:
I can allow more names, but the logic for getting the composition index, e.g. ` const unsigned int id_upper = this->introspection().compositional_index_for_name("upper");`, would get much more complicated as well.

naliboff replied:
You are right, this will get more complicated really quickly. More thoughts on this in the review summary.

naliboff commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,257 @@
+      AssertThrow(this->introspection().compositional_name_exists("lower"),ExcMessage("We need a compositional field called 'lower' representing the lower crust."));
+      AssertThrow(this->introspection().compositional_name_exists("mantle_L"),ExcMessage("We need a compositional field called 'mantle_L' representing the lithospheric part of the mantle."));
+
+      // For now, we assume a 3-layer system with an upper crust, lower crust and lithospheric mantle
```
Would it be possible to generalize the plugin to have an infinite number of layers within the lithosphere? I'll go back and look at the Chapman equations to see if this is possible.

naliboff replied:
The equations can also be reformulated such that the surface heat flow and temperature are provided, but the basal temperature is not specified. However, further generalization may be beyond the scope here.

naliboff commented on tests/continental_extension.prm:
```diff
@@ -0,0 +1,246 @@
+
+#### Parameters describing the model
+
+# Model geometry (400x100 km, 2 km spacing)
```
10 km spacing here

naliboff commented on tests/continental_extension.prm:
```diff
@@ -0,0 +1,246 @@
+#### 
```
You could remove a good chunk of the parameter file by using:
`include $ASPECT_SOURCE_DIR/cookbooks/continental_extension.prm`

naliboff commented on tests/continental_extension.prm:
```diff
@@ -0,0 +1,246 @@
+end
+
+# Mesh refinement specifications (no mesh refinement, 
+# but the coarse mesh is already 200x50, see above)
```
`mesh is already 40x10`

naliboff commented on tests/continental_extension.prm:
```diff
@@ -0,0 +1,246 @@
+  end
+end
+
+# Initial temperature field
```
Can you remove this documentation and the `function` section?

anne-glerum commented:
Hi @naliboff I addressed/answered your comments.

anne-glerum commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,276 @@
+          densities.push_back(temp_densities[id_mantle_L+1]);
+
+          // Thermal diffusivity kappa = k/(rho*cp), so thermal conducitivity k = kappa*rho*cp
+          conductivities.push_back(temp_thermal_diffusivities[id_upper+1] * densities[0] * temp_heat_capacities[id_upper+1]);
```
It's correct. The densities have already been put in the right order in lines 210-212 by selecting the needed densities from `temp_densities`. The diffusivities still need to be selected with the ids of the compositional fields however.

anne-glerum commented:
I'm not attached to the names of the fields, I just stuck to those in the cookbook. I've changed them to what you suggested and addressed the rest of the comments.

gassmoeller commented on include/aspect/initial_temperature/continental_geotherm.h:
```diff
@@ -0,0 +1,116 @@
+
+        /**
+         * Vector for the thicknesses of the compositional fields
+         * representing the layers `upper`, `lower` and `mantle_L`.
```
Please update names here.

gassmoeller commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,269 @@
+      // Retrieve the indices of the fields that represent the lithospheric layers.
+      AssertThrow(this->introspection().compositional_name_exists("upper_crust"),ExcMessage("We need a compositional field called 'upper' representing the upper crust."));
+      AssertThrow(this->introspection().compositional_name_exists("lower_crust"),ExcMessage("We need a compositional field called 'lower' representing the lower crust."));
+      AssertThrow(this->introspection().compositional_name_exists("lithospheric_mantle"),ExcMessage("We need a compositional field called 'mantle_L' representing the lithospheric part of the mantle."));
```
The three Asserts need updates to the names

gassmoeller commented on source/initial_temperature/continental_geotherm.cc:
```diff
@@ -0,0 +1,269 @@
+                                              "even though in principle the heat conduction equation can be solved "
+                                              "for any number of layers. The naming of the compositional fields "
+                                              "that represent the layers is also very specific, namely 'upper', "
+                                              "'lower', and 'mantle_L'. "
```
Need to update names.

anne-glerum commented:
@gassmoeller done :)

gassmoeller commented:
Great thanks!
